core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/source/CodeFormatter.java

```diff
@@ -0,0 +1,75 @@
+/**
+ * AET
+ *
+ * Copyright (C) 2018 Cognifide Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.cognifide.aet.job.common.comparators.source;
+
+import com.cognifide.aet.job.common.comparators.source.visitors.ContentVisitor;
+import com.cognifide.aet.job.common.comparators.source.visitors.MarkupVisitor;
+import com.cognifide.aet.job.common.comparators.source.visitors.NodeTraversor;
+import org.apache.commons.lang3.StringUtils;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+
+class CodeFormatter {
+
+  String format(String code, SourceCompareType sourceCompareType) {
+    String result;
+    switch (sourceCompareType) {
+      case MARKUP:
+        result = formatCodeMarkup(code);
+        break;
+      case CONTENT:
+        result = formatCodeContent(code);
+        break;
+      case ALLFORMATTED:
+        result = formatCodeAllFormatted(code);
+        break;
+      default:
+        result = code;
+        break;
+    }
+    return result;
+  }
+
+  private String formatCodeMarkup(String code) {
+    Document doc = Jsoup.parse(code);
+    NodeTraversor traversor = new NodeTraversor(new MarkupVisitor());
+    traversor.traverse(doc);
+    return doc.html();
+  }
+
+  private String formatCodeContent(String code) {
+    Document doc = Jsoup.parse(code);
+    ContentVisitor visitor = new ContentVisitor();
+    NodeTraversor traversor = new NodeTraversor(visitor);
+    traversor.traverse(doc);
+    return visitor.getFormattedText();
+  }
+
+  private String formatCodeAllFormatted(String code) {
+    Document doc = Jsoup.parse(code);
+    return removeEmptyLines(doc.outerHtml());
+  }
+
+  // package scoped for unit test
+  String removeEmptyLines(String source) {
+    String result = source;
+    if (StringUtils.isNotBlank(source)) {
+      result = result.replaceAll("(?m)^[ \t]*[\r\n]+", "");
```

Comment:
Instead of using regexp, could we try with [output settings escape mode](https://stackoverflow.com/a/39238900) ?
In my opinion regular expressions are error-prone...
If not, could you please use `Pattern.MULTILINE` ? and also `\s` for whitespace characters?

Reply:
Done, thx